cub/cub/device/dispatch/kernels/reduce.cuh

```diff
@@ -248,7 +260,7 @@ CUB_DETAIL_KERNEL_ATTRIBUTES __launch_bounds__(
   {
     if (threadIdx.x == 0)
     {
-      *d_out = init;
+      *d_out = detail::reduce::unwrap_empty_problem_init(init);
```

Comment:
```suggestion
      *d_out = static_cast<InitT>(init);
```
Should be enough to trigger `operator T()` of `empty_problem_init_t`. You can then remove the two overloads of `unwrap_empty_problem_init`.

Reply:
The implicit cast doesn't work under certain scenarios, e.g., when using a fancy iterator, such as the `transform_output_iterator` taking a template argument. This issue was revealed when using fancy iterators in the streaming implementation.